File: src/main/java/com/saomc/resources/StringNames.java
package com.saomc.resources;

import com.saomc.SAOCore;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.text.translation.I18n;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public final class StringNames {

    public static final ResourceLocation gui = new ResourceLocation(SAOCore.MODID, "textures/gui.png");
    public static final ResourceLocation guiCustom = new ResourceLocation(SAOCore.MODID, "textures/guiCustom.png");
    public static final ResourceLocation slot = new ResourceLocation(SAOCore.MODID, "textures/slot.png");
    public static final ResourceLocation icons = new ResourceLocation(SAOCore.MODID, "textures/icons.png");
    public static final ResourceLocation iconsCustom = new ResourceLocation(SAOCore.MODID, "textures/iconsCustom.png");
    public static final ResourceLocation effects = new ResourceLocation(SAOCore.MODID, "textures/gui.png");
    public static final ResourceLocation effectsCustom = new ResourceLocation(SAOCore.MODID, "textures/guiCustom.png");
    public static final ResourceLocation entities = new ResourceLocation(SAOCore.MODID, "textures/entities.png");
    public static final ResourceLocation entitiesCustom = new ResourceLocation(SAOCore.MODID, "textures/entitiesCustom.png");
    public static final ResourceLocation particleLarge = new ResourceLocation(SAOCore.MODID, "textures/particleLarge.png");
    public static final String FRIEND_REQUEST_TITLE = I18n.translateToLocal("guiFriendRequestTitle");
    public static final String FRIEND_REQUEST_TEXT = "guiFriendRequestText";
    public static final String PARTY_INVITATION_TITLE = I18n.translateToLocal("guiPartyInviteTitle");
    public static final String PARTY_INVITATION_TEXT = "guiPartyInviteText";
    public static final String PARTY_DISSOLVING_TITLE = I18n.translateToLocal("guiPartyDissolvingTitle");
    public static final String PARTY_DISSOLVING_TEXT = I18n.translateToLocal("guiPartyDissolvingText");
    public static final String PARTY_LEAVING_TITLE = I18n.translateToLocal("guiPartyLeaveTitle");
    public static final String PARTY_LEAVING_TEXT = I18n.translateToLocal("guiPartyLeaveText");
    public static final String MESSAGE_TITLE = I18n.translateToLocal("guiMessageTitle");
    public static final String MESSAGE_FROM = "guiMessageFrom";
    public static final String DEAD_ALERT = I18n.translateToLocal("guiDeadAlert");

    private StringNames() {
    }

}

File: src/main/java/com/tencao/saoui/resources/StringNames.kt
/*
 * Copyright (C) 2016-2019 Arnaud 'Bluexin' Solé
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tencao.saoui.resources

import com.tencao.saoui.GLCore
import com.tencao.saoui.SAOCore
import com.tencao.saoui.effects.StatusEffects
import com.tencao.saoui.themes.ThemeManager
import com.tencao.saoui.util.append
import net.minecraft.util.ResourceLocation
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

@SideOnly(Side.CLIENT)
object StringNames {

    private fun logMissingAndUse(type: String, default: ResourceLocation): ResourceLocation {
        SAOCore.LOGGER.info("Theme {} missing custom {}, defaulting to SAO", ThemeManager.currentTheme, type)
        return default
    }

    fun init() {
        val textureRoot = ThemeManager.currentTheme.texturesRoot

        gui = GLCore.useTextureIfExists(textureRoot.append("gui.png"))
            ?: logMissingAndUse("gui", defaultGui)
        slot = GLCore.useTextureIfExists(textureRoot.append("slot.png"))
            ?: logMissingAndUse("slot", defaultSlot)
        entities = GLCore.useTextureIfExists(textureRoot.append("entities.png"))
            ?: logMissingAndUse("entity health bars", defaultEntities)
        particleLarge = GLCore.useTextureIfExists(textureRoot.append("particlelarge.png"))
            ?: logMissingAndUse("death particles", defaultParticleLarge)

        val missingEffects = StatusEffects.values().filter {
            !GLCore.checkTexture(textureRoot.append("status_icons/${it.name.toLowerCase()}.png"))
        }
        statusIcons = if (missingEffects.isEmpty()) {
            textureRoot.append("status_icons/")
        } else logMissingAndUse(
            "status icons ${missingEffects.map(StatusEffects::name)}",
            defaultStatusIcons
        )
    }

    lateinit var gui: ResourceLocation
    lateinit var slot: ResourceLocation
    lateinit var entities: ResourceLocation
    lateinit var particleLarge: ResourceLocation
    lateinit var statusIcons: ResourceLocation

    private val defaultGui = ResourceLocation(SAOCore.MODID, "textures/sao/gui.png")
    private val defaultSlot = ResourceLocation(SAOCore.MODID, "textures/slot.png")
    private val defaultEntities = ResourceLocation(SAOCore.MODID, "textures/sao/entities.png")
    private val defaultParticleLarge = ResourceLocation(SAOCore.MODID, "textures/sao/particlelarge.png")
    private val defaultStatusIcons = ResourceLocation(SAOCore.MODID, "textures/sao/status_icons/")
}
